Tasks/Gradle/CodeAnalysis/Common/CodeAnalysisResultPublisher.ts

```diff
@@ -0,0 +1,144 @@
+            return groups[group];
+        });
+    }
+
+    private uploadMdSummary(content: string) {
+        var buildSummaryFilePath: string = path.join(this.stagingDir, 'CodeAnalysisBuildSummary.md');
+        tl.mkdirP(this.stagingDir);
+        fs.writeFileSync(buildSummaryFilePath, content);
+
+        tl.debug('[CA] Uploading build summary from ' + buildSummaryFilePath);
+
+        tl.command('task.addattachment', {
+            'type': 'Distributedtask.Core.Summary',
+            'name': tl.loc('codeAnalysisBuildSummaryTitle')
+        }, buildSummaryFilePath);
+    }
+
+    private createSummaryContent(): string {
+
+        var buildSummaryLines: string[] = [];
+        var resultsGroupedByTool: AnalysisResult[][] = this.groupBy(this.analysisResults, (o: AnalysisResult) => { return o.toolName; });
+
+        for (var resultGroup of resultsGroupedByTool) {
+            var summaryLine = this.createSummaryLine(resultGroup);
+            buildSummaryLines.push(summaryLine);
+        }
+
+        tl.debug(`[CA] Build Summary: ${buildSummaryLines}`);
+
+        if (buildSummaryLines.length > 0) {
+            buildSummaryLines.push('');
+            buildSummaryLines.push('Code analysis results can be found in the \'Artifacts\' tab.');
+        }
+
+        return buildSummaryLines.join('  \r\n');
+    }
+
+    // For a given code analysis tool, create a one-line summary from multiple AnalysisResult objects.
+    private createSummaryLine(analysisResultsGroup: AnalysisResult[]): string {
+        var violationCount: number = 0;
+        var affectedFileCount: number = 0;
+        var toolName = analysisResultsGroup[0].toolName;
+
+        analysisResultsGroup.forEach((analysisResult: AnalysisResult) => {
+            violationCount += analysisResult.violationCount;
+            affectedFileCount += analysisResult.affectedFileCount;
+        });
+
+        if (violationCount > 1) {
+            if (affectedFileCount > 1) {
+                // Looks like: 'PMD found 13 violations in 4 files.'
+                return tl.loc('codeAnalysisBuildSummaryLine_SomeViolationsSomeFiles', toolName, violationCount, affectedFileCount);
+            }
+            if (affectedFileCount === 1) {
+                // Looks like: 'PMD found 13 violations in 1 file.'
+                return tl.loc('codeAnalysisBuildSummaryLine_SomeViolationsOneFile', toolName, violationCount);
+            }
+        }
+        if (violationCount === 1 && affectedFileCount === 1) {
+            // Looks like: 'PMD found 1 violation in 1 file.'
+            return tl.loc('codeAnalysisBuildSummaryLine_OneViolationOneFile', toolName);
+        }
+        if (violationCount === 0) {
+            // Looks like: 'PMD found no violations.'
+            return tl.loc('codeAnalysisBuildSummaryLine_NoViolations', toolName);
+        }
+
+        // There should be no valid code reason to reach this point - '1 violation in 4 files' is not expected
```

Comment:
It would probably be neater and easier to read if you used a _switch_ statement rather than multiple _if_ - _return_ statements. That would also put the error case code closer to where the error occurred.